src/libsyntax/ext/expand.rs

```diff
@@ -447,6 +471,30 @@ fn expand_annotatable(mut item: Annotatable, fld: &mut MacroExpander) -> SmallVe
 
             let modified = match *extension {
                 MultiModifier(ref mac) => mac.expand(fld.cx, attr.span, &attr.node.value, item),
+                SyntaxExtension::AttrProcMacro(ref mac) => {
+                    let attr_toks = TokenStream::from_tts(span_to_tts(&fld.cx.parse_sess,
+                                                                      attr.span));
+                    let item_toks = TokenStream::from_tts(span_to_tts(&fld.cx.parse_sess,
+                                                                      item.span()));
+                    let result = mac.expand(fld.cx, attr.span, attr_toks, item_toks);
+                    match item {
+                        Annotatable::Item(_) => result.make_items()
+                                                      .unwrap_or(SmallVector::zero())
```

Comment:
`MultiModifier`s don't have this problem since they return `Annotatable`s instead of `MacResult`s. If a `MultiModifier` returns the wrong kind of `Annotatable`, we currently panic.

Consumers of a `MacResult` always error on `None` in any position.
If the a `make_*` method reports an error, it returns `Some(MacroGenerable::dummy(span))`.